src/linux_native.rs

```diff
@@ -0,0 +1,670 @@
+    {
+        Some(Ok(s)) => s,
+        None | Some(Err(_)) => return None,
+    };
+
+    // Thus far we've gathered all the common attributes.
+    let info = DeviceInfo {
+        path,
+        vendor_id: vid,
+        product_id: pid,
+        serial_number: osstring_to_string(serial.into()),
+        release_number: 0,
+        manufacturer_string: WcharString::None,
+        product_string: WcharString::None,
+        usage_page: 0,
+        usage: 0,
+        interface_number: -1,
+        bus_type,
+    };
+
+    // USB has a bunch more information but everything else gets the same empty
+    // manufacturer and the product we read from the property above.
+    let info = match bus_type {
+        BusType::Usb => fill_in_usb(raw_device, info, name),
+        _ => DeviceInfo {
+            manufacturer_string: WcharString::String("".into()),
+            product_string: osstring_to_string(name.into()),
+            ..info
+        },
+    };
+
+    if let Ok(descriptor) = HidrawReportDescriptor::from_syspath(raw_device.syspath()) {
+        let mut usages = descriptor.usages();
+
+        // Get the first usage page and usage for our current DeviceInfo
+        if let Some((usage_page, usage)) = usages.next() {
+            infos.push(DeviceInfo {
+                usage_page,
+                usage,
+                ..info
+            });
+
+            // Now we can create DeviceInfo for all the other usages
+            for (usage_page, usage) in usages {
+                let prev = infos.last().unwrap();
+
+                infos.push(DeviceInfo {
+                    usage_page,
+                    usage,
+                    ..prev.clone()
+                })
+            }
+        }
+    } else {
+        infos.push(info);
+    }
+
+    Some(infos)
+}
+
+/// Fill in the extra information that's available for a USB device.
+fn fill_in_usb(device: &udev::Device, info: DeviceInfo, name: &OsStr) -> DeviceInfo {
+    let usb_dev = match device.parent_with_subsystem_devtype("usb", "usb_device") {
+        Ok(Some(dev)) => dev,
+        Ok(None) | Err(_) => {
+            return DeviceInfo {
+                manufacturer_string: WcharString::String("".into()),
+                product_string: osstring_to_string(name.into()),
+                ..info
+            }
+        }
+    };
+    let manufacturer_string = attribute_as_wchar(&usb_dev, "manufacturer");
+    let product_string = attribute_as_wchar(&usb_dev, "product");
+    let release_number = attribute_as_u16(&usb_dev, "bcdDevice");
+    let interface_number = device
+        .parent_with_subsystem_devtype("usb", "usb_interface")
+        .ok()
+        .flatten()
+        .map(|ref dev| attribute_as_i32(dev, "bInterfaceNumber"))
+        .unwrap_or(-1);
+
+    DeviceInfo {
+        release_number,
+        manufacturer_string,
+        product_string,
+        interface_number,
+        ..info
+    }
+}
+
+// From linux/hidraw.h
+struct HidrawReportDescriptor {
```

Comment:
Why not simply store this in a vector or boxed slice?
You can still newtype HidrawReportDescriptor around it.

Reply:
It's just what the C version uses, which is not what we would do fresh in Rust. We probably could thingly-wrap a `Vec<u8>` or something.

Reply:
This is now just wrapping a vector, so you could have descriptors larger than the kernel allows, but that's probably fine since the kernel is giving us the descriptor.